pkg/hive/cell/metric.go

```diff
@@ -0,0 +1,139 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cell
+
+import (
+	"fmt"
+	"reflect"
+	"runtime"
+
+	"go.uber.org/dig"
+
+	"github.com/cilium/cilium/pkg/hive/internal"
+	pkgmetric "github.com/cilium/cilium/pkg/metrics/metric"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func Metric[S any](ctor func() S) Cell {
+	caller := "caller"
+	if _, file, no, ok := runtime.Caller(1); ok {
+		caller = fmt.Sprintf("%s:%d", file, no)
+	}
+
+	var nilOut S
+	outTyp := reflect.TypeOf(nilOut)
+	if outTyp.Kind() == reflect.Ptr {
+		outTyp = outTyp.Elem()
+	}
+
+	if outTyp.Kind() != reflect.Struct {
+		panic(fmt.Sprintf(
+			"cell.Metric must be invoked with a constructor function that returns a struct or pointer to a struct, "+
+				"%s supplied a constructor which returns a %d",
+			caller,
+			outTyp.Kind(),
+		))
+	}
+
+	// Lets be strict for now, could lift this in the future if we ever need to
+	if outTyp.NumField() == 0 {
+		panic(fmt.Sprintf(
+			"cell.Metric must be invoked with a constructor function that returns exactly a struct with at least 1 "+
+				"metric, %s supplied a constructor which returns a struct with zero fields",
+			caller,
+		))
+	}
+
+	var withMeta pkgmetric.WithMetadata
+	withMetaTyp := reflect.TypeOf(&withMeta).Elem()
+
+	var collector prometheus.Collector
+	collectorTyp := reflect.TypeOf(&collector).Elem()
+	for i := 0; i < outTyp.NumField(); i++ {
+		field := outTyp.Field(i)
+		if !field.IsExported() {
```

Comment:
Why do we need this constraint on metrics, I don't think other cells enforce this?

Reply:
If a field containing a metric isn't exported, we can't make it separately available in Hive. That would result in the odd behavior where the package defining the metric struct can still access, modify and use the metric but it will never be exported. That would happen silently, so by enforcing this I hope to avoid users making this hard to find bug and easy to overlook type of bug.